imap_processing/glows/l1/glows_l1a_data.py

```diff
@@ -0,0 +1,411 @@
+                general_data_subset[prev_byte : prev_byte + item[1]], "big"
+            )
+            prev_byte = prev_byte + item[1]
+
+        return data_every_second
+        # data every second has padding, so it is exactly 40 bytes
+
+    def _generate_direct_events(self, direct_events: bytearray):
+        """Generate the list of direct events from the raw bytearray.
+
+        First, the starting timestamp is created from the first 8 bytes in the direct
+        event array. Then, the remaining events are processed based on a marker in the
+        first two bits of each section. If the marker is 0, it is uncompressed, and
+        the event is processed from the following 7 bytes. If it is 1, it is compressed
+        to two bytes, and if it is 2, the direct event is compressed to 3 bytes.
+
+        Parameters
+        ----------
+        direct_events: bytearray
+            bytearray containing direct event data
+
+        Returns
+        -------
+        processed_events: list[DirectEvent]
+            An array containing DirectEvent objects
+
+        """
+        # read the first direct event, which is always uncompressed
+        current_event = self._build_uncompressed_event(direct_events[:8])
+        processed_events = [current_event]
+
+        i = 8
+        while i < len(direct_events) - 1:
+            first_byte = int(direct_events[i])
+            i += 1
+            # Remove first two bits, which are used to mark compression
+            oldest_diff = first_byte & 0x3F
+            marker = first_byte >> 6
+
+            if (
+                marker == 0x0 and i < len(direct_events) - 7
+            ):  # uncompressed time stamp (8-bytes)
+                rest_bytes = direct_events[i : i + 7]
+                i += 7
+                part = bytearray([oldest_diff])
+                part.extend(rest_bytes)
+                current_event = self._build_uncompressed_event(part)
+
+            elif (
+                marker == 0x2 and i < len(direct_events) - 1
+            ):  # 2-byte compression of timedelta
+                current_event = self._build_compressed_event(
+                    direct_events[i : i + 2], oldest_diff, current_event.timestamp
+                )
+                i += 2
+
+            elif (
+                marker == 0x3 and i < len(direct_events) - 2
+            ):  # 3-byte compression of timedelta
+                current_event = self._build_compressed_event(
+                    direct_events[i : i + 3], oldest_diff, current_event.timestamp
+                )
+                i += 3
+
+            else:  # wrong-marker or hitting-the-buffer-end case
+                raise Exception(
+                    "error: unexpected marker %d or end of "
+                    "data stream %d %d" % (marker, i, len(direct_events))
+                )
+
+            # sanity check
+            assert current_event.timestamp.subseconds < GlowsConstants.SUBSECOND_LIMIT
```

Comment:
Since we are doing check/raises elsewhere do we want to continue that here as well?
```python
if not ...
   raise ValueError("Subseconds received are less than the GLOWS subsecond limit")
```